// ui-project-base/build-tools/build-static/rollup.config.js
const generatePackageJson = require('rollup-plugin-generate-package-json');
const mainPackageJson = require('../../package.json');
const commonjs = require('@rollup/plugin-commonjs');
const copy = require('rollup-plugin-copy');
const resolve = require('@rollup/plugin-node-resolve');

// This configures the rollup bundler to bundle only fonts, css & hbs helpers
// This generates a 'sub package' called sdc-project-base-static in a dist folder
const config = [
  {
    input: 'js/helpers/all.js',
    output: {
      file: 'dist/all.js',
      format: 'cjs'
    },
    plugins: [
        resolve(),
        commonjs(),
        generatePackageJson({
            baseContents: {
            name: 'ui-project-base-static',
            version: mainPackageJson.version,
            publishConfig: {
                "access": "public",
                "registry": "https://sky.jfrog.io/sky/api/npm/npm/"
            },
            }
        }),
        copy({
            targets: [
            { src: 'build-tools/build-static/.npmrc', dest: 'dist' },
            { src: 'fonts/*', dest: 'dist/fonts' },
            { src: 'css/base/*', dest: 'dist/css/base'}
            ]
        })
    ]
  },
];

module.exports = config
